src/Git/Index/Update.hs

```diff
@@ -60,6 +69,41 @@ updateIndex relpath = do
       Normal ie -> Just ie
       _ -> Nothing
 
+parseContent :: MonadFail m => Parser a -> Path.AbsFile -> IO (m a)
+parseContent parser path = do
+    content <- openBinaryFile path ReadMode >>= LBS.hGetContents
+    pure $ lazyParseOnly (parser <* endOfInput) content
+
+getIgnores :: (MonadIO m, MonadError GitError m) => Path.AbsDir -> m Ignores
+getIgnores dirP = do
+    eMfIgnores <- liftIO $ (
+        try $ parseContent ignoresP $ dirP </> Path.relFile ".gitignore"
+        :: IO (Either IOException (Either String Ignores)))
+    case eMfIgnores of
+        Left _ -> pure mempty  -- Failed to open .gitignore
+        Right mfIgnores -> either (throwError . ParseError) pure mfIgnores
+
+updateIndexDir
+  :: (MonadIO m, MonadError GitError m, MonadState Index m, MonadReader Repo m)
+  => Path.RelDir -> m ()
+updateIndexDir targetDir = do
+    repoRoot <- repoFilePath <$> ask
+    parentIgnores <- mapM getIgnores $ parentDirPaths repoRoot
+    addDirContents repoRoot (foldr (<>) mempty parentIgnores) targetDir
+  where
+    (_, targetSegs, _) = Path.splitPath $ Path.normalise targetDir
```

Comment:
Ah, interesting. `Path.splitPath` is what I couldn't find before. Should we be normalising here?

Reply:
I think so, otherwise for dir paths like `bar/foo/` we get `["bar", "foo"]` instead of `["bar"]`.

Reply:
I'm confused - the [docs for normalise](https://hackage.haskell.org/package/pathtype-0.8.1/docs/System-Path-Generic.html#v:normalise) just mention eliminating intermediary `.`s from the path. How and why does it drop the `foo`?

Reply:
Ah, sorry, yes, it doesn't give `.` as a parent seg of `./` (which would load the same gitignore twice). I was presumably thinking of something else.